separatedunit.cpp

```diff
@@ -0,0 +1,1286 @@
+    bitLenInt i, j;
+    QbLookup qbl;
+    QbListEntry qbe;
+    for (i = 0; i < length; i++) {
+        qbl = qubitLookup[start + i];
+        qbe.cu = qbl.cu;
+        qbe.start = qbl.qb;
+        qbe.length = 1;
+        qbList[i] = qbe;
+    }
+    // The ordering of bits returned is unimportant, so we can better optimize by sorting this list by CoherentUnit
+    // index and qubit index, to maximize the reduction of the list.
+    std::sort(qbList.begin(), qbList.end(), compare);
+    // If contiguous sublists in the list we just sorted are also contiguous in the same coherent unit, we can combine
+    // them to optimize with register-wise gate methods.
+    j = 0;
+    for (i = 0; i < (length - 1); i++) {
+        if ((qbList[j].cu == qbList[j + 1].cu) && ((qbList[j].start + qbList[j].length) == qbList[j + 1].start)) {
+            qbList[j].length += qbList[j + 1].length;
+            qbList.erase(qbList.begin() + j + 1);
+        } else {
+            j++;
+        }
+    }
+}
+
+/// Combines two lists returned by GetParallelBitList() by the same logic as that algorithm
+void SeparatedUnit::OptimizeParallelBitList(std::vector<QbListEntry>& qbList)
+{
+    if (qbList.size() < 2) {
+        return;
+    }
+
+    bitLenInt i, j;
+    bitLenInt length = qbList.size();
+    // The ordering of bits returned is unimportant, so we can better optimize by sorting this list by CoherentUnit
+    // index and qubit index, to maximize the reduction of the list.
+    std::sort(qbList.begin(), qbList.end(), compare);
+    // If contiguous sublists in the list we just sorted are also contiguous in the same coherent unit, we can combine
+    // them to optimize with register-wise gate methods.
+    j = 0;
+    for (i = 0; i < (length - 1); i++) {
+        if ((qbList[j].cu == qbList[j + 1].cu) && ((qbList[j].start + qbList[j].length) == qbList[j + 1].start)) {
+            qbList[j].length += qbList[j + 1].length;
+            qbList.erase(qbList.begin() + j + 1);
+        } else {
+            j++;
+        }
+    }
+}
+
+/// Entangle and sort the indices of a list of CoherentUnit objects
+void SeparatedUnit::EntangleBitList(std::vector<QbListEntry> qbList)
+{
+    if (qbList.size() < 2) {
+        return;
+    }
+
+    bitLenInt i, j, k;
+    bitLenInt firstCu, cuLen, invLookup, cuRemoved;
+    QbListEntry qbe;
+
+    firstCu = qbList[0].cu;
+    k = coherentUnits[firstCu]->GetQubitCount();
+    for (i = 1; i < qbList.size(); i++) {
+        qbe = qbList[i];
+        cuLen = coherentUnits[qbe.cu]->GetQubitCount();
+        for (j = 0; j < cuLen; j++) {
+            invLookup = qubitInverseLookup[qbe.cu * qubitCount + j];
+            qubitLookup[invLookup].cu = firstCu;
+            qubitLookup[invLookup].qb = k + j;
+            qubitInverseLookup[firstCu * qubitCount + k + j] = invLookup;
+        }
+        coherentUnits[firstCu]->Cohere(*(coherentUnits[qbe.cu]));
```

Comment:
Seems particularly expensive to Cohere one CU at a time.  Instead, perhaps create a new CU that takes an arbitrary set of CUs to cohere from at a single go?  Substantially reducing the number of allocations, for example.

Reply:
I was going to go the route of combining them smallest to largest in iterations, but this might be even better. I'm going to try to implement this, so I'll comment here again on the result.